### ISSUE TITLE
24 hour spontaneous update doesn't seem specced

### ISSUE THREAD

**Issue by mattto** [slightlyoff/ServiceWorker#514]
A few points related to the 24 hour update:

* There's a note that: "update() pings the server for an updated version of this script without consulting caches. This is conceptually the same operation that UA does maximum once per every 24 hours."
But I don't see where this operation is defined.
* The "without consulting caches" part seems wrong. update() leads to Update, which busts the browser cache only once every 24 hours.
* The Soft Update algorithm mentions: "The user agent may call this as often as it likes to check for updates." This seems to contradict the "maximum once every 24 hours" statement (maybe that is meant to be "at least once..."?)
* I realize that Fetch schedules a Soft Update, and Update busts the browser cache every 24 hours. But that's different than a spontaneous 24 hour update, no?
* nit: It might be more consistent and precise to change the "forcing a network fetch if cached entry is greater than 1 day old" to "greater than 24 hours old".

**Comment by jakearchibald** [comment]
We should store the max age along with the SW, but capped to 24hrs. When we fetch a SW as part of an update, we decide which cache mode to use based on this max age. This means we don't have to look at the HTTP cache before we fetch the SW.

**Comment by jakearchibald** [comment]
Additionally, we should handle updates like this:

* When we fire an event in the SW, we should trigger a soft update if we haven't done one within 24hrs
* The fetch event is the exception, which triggers an update every navigation

**Comment by jungkees** [comment]
>A call to update() should cause the server to get hit even if there's currently a SW in the installing state.

Decided to remove update() method's step 3 to address it.

**Comment by jungkees** [comment]
Addressed those points from f2f: 8ea4c1e2f0ebb612962fffed4a91ee21a7b9df00. Let me close it after having feedback.

**Comment by jungkees** [comment]
Thanks for review @slightlyoff! Closing.

**Comment by mattto** [comment]
I'm surprised that update() no longer forces bypassing the HTTP cache as https://code.google.com/p/chromium/issues/detail?id=520068 is saying. I don't recall deciding that at F2F? What was the rationale?

I would think if an author writes update(), they expect to ping the server for the latest script.

**Comment by jungkees** [comment]
I'm not entire sure we've talked about exactly .update() also applies the same rule, but checked with @jakearchibald it seems fine to apply the same rule to all the updates gated on Update algorithm. Now devs can adjust the cap by setting the Cache-Control max-age header themselves, and they could control this when an immediate update is needed. The reason we changed it this way is mostly consistency among the different update requests. `.update({ cache-mode: "reload" }) // or "default"` seems to be another way to handle this? WDYT?

**Comment by mattto** [comment]
Relying on setting max age doesn't sound right. You may have already set it to 100000, so by the time you realize you want it to set it to 0 to push out an emergency update it's too late.

I guess cache-mode param in update() is an OK solution... seems like unnecessary complexity to me though.

**Comment by jungkees** [comment]
I think I'm fine with that. We might want to consider `.update()` as a method that devs can explicitly ping the network resource anytime. Let's also forget about the cache-mode option. 

I believe `.update()` is the only caller of the Update algorithm that'd always bypass cache among all the update paths for now. Do you agree?

I'm planning to add an optional force bypass cache flag in Soft Update and Update algorithms.

**Comment by mattto** [comment]
Sounds right, maybe @jakearchibald can jump in to confirm.

**Comment by jungkees** [comment]
Yeah, in reflection, exposing an option bag would come in handy for devs depending on what they want to do. /cc @jakearchibald

**Comment by jungkees** [comment]
Added _force bypass cache flag_ which can be needed in any future work: 3d675eb950d3d546f4ac15ceecca2f4a15cc34f5.

**Comment by jakearchibald** [comment]
`.update()` allows me to check for updates to the service worker outside of the usual schedule, but I don't see why the SW fetch should behave any differently to `fetch(url)` or `cache.add(url)` in terms of cache interaction.

I might call `.update()` when page visibility changes, or on a long `setTimeout`, but if I don't want it to http-cache it shouldn't have a `max-age` higher than 0 set.

> You may have already set it to 100000, so by the time you realize you want it to set it to 0 to push out an emergency update it's too late

If all your resources are served from the cache, all you can do is fix the SW cache headers and wait (a maximum of) 24hrs. Even if `.update` ignores the cache, it's impossible to get the page to run that.

If you're not serving everything from the cache, you can call `.register` and give the SW url a querystring, eg 'sw.js?v2'. Although I'm not against an option to `.update` that causes a cache-bypass.

If I'm trying to recover from a max-age incident, would I call `.update({bypassCache: true})` per page load for 24hrs? Will that cause double requests for users that didn't see a SW with a max-age?

**Comment by mattto** [comment]
OK Jake's explanation makes sense. I just wanted to raise the alarm since the spec long had the non-normative note "update() pings the server for an updated version of this script without consulting caches" and I didn't recall overturning that in the F2F. Sounds like Jake et al have thought it through so I'm OK with it.

**Comment by jungkees** [comment]
Agreed. Let's stick to what has been discussed here.

**Comment by jungkees** [comment]
From https://codereview.chromium.org/1283273002/, another question came up about scheduled updates by the user agents.

Do we consider the service worker's max age set by min(Cache-Control: max-age's value, 86400) a desired lifetime of a service worker? In which case, using this value for an auto-update seems fine.

On the other hand, it still sounds reasonable to use 24h rule for auto-updating a stale SW as it was. Scheduling the updates for all the installed service workers in a short timeframe only depending on the Cache-Control max-age values would be too much for the user agents I guess.

If we want to keep the 24h auto-update rule, the service worker's max age internal slot in the spec is not necessary at all. As point out by @nhiroki (https://codereview.chromium.org/1283273002/diff/20001/content/browser/service_worker/service_worker_context_request_handler.cc#newcode72), the Cache-Control max-age will cover all the cases in the network layer.

**Comment by jakearchibald** [comment]
I'm struggling to get my head around this. What's the difference?

I don't think we want to force the browser to perform an update on an interval. Although the browser may check for updates more regularly, it only *has* to do it after navigation, or when other specs trigger updates (eg on push).

**Comment by jungkees** [comment]
I was talking about the "24 hour spontaneous update" that the ticket had started for. I think I was a bit confused about it as if the user agent had to do this spontaneous update regularly even in the service worker's max-age interval.. But it seems we don't want to specify that, right?

The current setting in the Soft Update algorithm should be sufficient I guess: "The user agent may call this as often as it likes to check for updates."

> it only has to do it after navigation, or when other specs trigger updates (eg on push).

Yes, this is correct. We're doing that.

**Comment by jakearchibald** [comment]
> as if the user agent had to do this spontaneous update regularly even in the service worker's max-age interval.. But it seems we don't want to specify that, right?

Agreed, we don't want to specify that.

**Comment by jungkees** [comment]
Re-open to discuss the criteria for bypassing the browser cache. 

The registration's _last update time_ is used for this purpose in the spec and the value is updated only when a new version is fetched from the network and the resource is new. (I.e., it doesn't care whether the subsequent installation fails. And it's not updated when the response's status was 304 - i.e. the resource in the server is still byte-for-byte match.)

On the other hand, the current chrome implementation show a bit different behavior. It has a corresponding member, |last_update_check|, on the registration, and the value is updated in two cases: 1) when the response's status was 304 or 2) the new version of the resource has been received and successfully installed. (That is, it doesn't update the value when the installation fails even if it actually fetched a new script resource from the network.)

I've been pondering upon this browser cache bypassing rule for a bit. I think the overall rule is "we must go to the network if we haven't within 24 hours". (max-age is covered by the network layer anyway.)

That considered, I'd want to suggest we just use the simplest rule where the registration's _last update check time_ is updated if and only if the response is gotten from the network. (i.e. including 304 case; and we don't care about the subsequent installation for this new resource.) I think we can check the condition using response's cache state is not "local".

I think we should not take the installation failure into account here as, coming back to the overall rule, we already hit the network and the browser cache should have already been bumped with this new resource.

Thoughts?

/cc @jakearchibald @nhiroki

**Comment by slightlyoff** [comment]
Install failure for a new version seems like a situation we might be able to handle independently; i.e. by retrying the install.

Regarding when we should update the timestamp, the simpler rule which doesn't pay attention to install success or failure doesn't strike me as being problematic. A site with a busted SW is the reason to have the backstop and potentially DoSing a server for which there's a failed SW install seems like a bad idea (the natural consequence of re-checking every navigation after 24 hours when the install fails).

**Comment by jungkees** [comment]
Addressed: 569863f3acd0f5d79beab9d8c63210b8286ee507. Now the _last update check time_ value is updated when it receives a valid response upon a request that bypassed the browser cache. (the Update algorithm step 4.13.) In addition, the service worker's max age internal slot has been removed as it's not required to specify the behavior. A note is added in the Update algorithm step 4.10 instead.

**Comment by jungkees** [comment]
Closing.

**Comment by wanderview** [comment]
Gecko bug to implement the new update mechanism: https://bugzilla.mozilla.org/show_bug.cgi?id=1207727

**Comment by jungkees** [comment]
Please note that there's an ongoing discussion in Blink: https://codereview.chromium.org/1283273002/. There's a possibility we'd have to move the time stamp updating step from the Update algorithm step 4.13 to 4.24 (i.e. from before validating the response to after the validation).

**Comment by jungkees** [comment]
Reopen it to address the issue encountered during the implementation as noted in the last comment.

**Comment by jungkees** [comment]
F2F: Agreed to move the step updating the timestamp from 4.13 to 4.24.
